### PR TITLE
IC-1289 add contracts for appointment APIs

### PR DESCRIPTION
**Pull request by tomsmyers**
## What does this pull request do?

adds contracts for creating, updating, getting, and deleting action plan appointments (sessions)

## What is the intent behind these changes?

implement the SP side of viewing appointments/sessions for an assigned referral

### CODE REVIEW

File: server/services/interventionsService.ts [hunk 1]
```diff
@@ -399,4 +405,41 @@ export default class InterventionsService {
...
+      path: `/action-plan/${actionPlanId}/appointment`,
+      headers: { Accept: 'application/json' },
+      data: { ...appointment },
+    })) as ActionPlanAppointment
+  }
+
+  async updateActionPlanAppointment(
+    token: string,
+    actionPlanId: string,
+    appointment: ActionPlanAppointment
```
Comment (mahalrmoj) [inline server/services/interventionsService.ts:433]: ActionPlanAppointment contains a sessionNumber. In the case of an update is it applicable?

Comment (mahalrmoj) [reply server/services/interventionsService.ts:433]: I do favour it as a parameter as it is part of the key to identifying an appointment and thus something that can't change, whereas the body contains elements that can change.

Comment (tomsmyers) [reply server/services/interventionsService.ts]: done

**Review by mahalrmoj — APPROVED** [review]

File: server/services/interventionsService.ts [hunk 2]
```diff
@@ -406,11 +406,24 @@ export default class InterventionsService {
...
+  async createActionPlanAppointment(
+    token: string,
+    actionPlanId: string,
+    appointment: ActionPlanAppointment
+  ): Promise<ActionPlanAppointment> {
+    const restClient = this.createRestClient(token)
+    return (await restClient.post({
+      path: `/action-plan/${actionPlanId}/appointment`,
+      headers: { Accept: 'application/json' },
+      data: { ...appointment },
```
Comment (lawrence-forooghian) [inline server/services/interventionsService.ts:431]: Why not just 

```suggestion
      data: appointment,
```

File: server/services/interventionsService.ts [hunk 3]
```diff
@@ -426,4 +431,18 @@ export default class InterventionsService {
...
+    token: string,
+    actionPlanId: string,
+    sessionNumber: number,
+    appointmentUpdate: Partial<ActionPlanAppointmentUpdate>
+  ): Promise<ActionPlanAppointment> {
+    const restClient = this.createRestClient(token)
+    return (await restClient.patch({
+      path: `/action-plan/${actionPlanId}/appointment/${sessionNumber}`,
+      headers: { Accept: 'application/json' },
+      data: { ...appointmentUpdate },
```
Comment (lawrence-forooghian) [inline server/services/interventionsService.ts:445]: Same question as before about the use of the spread operator here.

File: server/views/serviceProviderReferrals/showReferral.njk [hunk 4]
```diff
@@ -26,6 +26,7 @@
         <p class="govuk-body">This intervention is assigned to <strong>{{ presenter.text.assignedTo }}</strong>.</p>
 
         <form method="post" action="{{ presenter.createActionPlanFormAction }}">
+          <input type="hidden" name="_csrf" value="{{csrfToken}}">
```
Comment (lawrence-forooghian) [inline server/views/serviceProviderReferrals/showReferral.njk:29]: What does this have to do with the rest of the PR? Can we ditch it? I think it's in George's other draft PR anyway.

File: server/services/interventionsService.test.ts [hunk 5]
```diff
@@ -1552,17 +1552,17 @@ pactWith({ consumer: 'Interventions UI', provider: 'Interventions Service' }, pr
     const actionPlanAppointments = [
       {
         sessionNumber: 1,
-        appointmentTime: '2021-05-13T12:30:000000Z',
```
Comment (lawrence-forooghian) [inline server/services/interventionsService.test.ts]: These are all modifications to lines originally introduced by this PR, right? If so can we please incorporate them into the commits where the lines were introduced?

File: server/services/interventionsService.test.ts [hunk 6]
```diff
@@ -1552,17 +1552,17 @@ pactWith({ consumer: 'Interventions UI', provider: 'Interventions Service' }, pr
     const actionPlanAppointments = [
       {
         sessionNumber: 1,
-        appointmentTime: '2021-05-13T12:30:00.000000Z',
+        appointmentTime: '2021-05-13T13:30:00+01:00',
```
Comment (lawrence-forooghian) [inline server/services/interventionsService.test.ts:1555]: I'm a little confused about this change. What are we trying to communicate with this time zone? Does the time zone passed here have some meaning, and if so what is the frontend meant to pass?

Comment (mahalrmoj) [reply server/services/interventionsService.test.ts:1555]: When implementing the contract tests server side a couple things came to light:

1) appointmentTime is in a period that is subject to BST (based on the local my server) and the serialisation library considers this when doing the serialisation, so whilst the time is materially the same, it serialises to a different string
2) if the milli's is set to zero, the component is omitted from the serialised form 

These changes were necessary to get the contract tests to function from the server side as it appears the verification is based on string values.

Comment (lawrence-forooghian) [reply server/services/interventionsService.test.ts:1555]: > appointmentTime is in a period that is subject to BST (based on the local my server) and the serialisation library considers this when doing the serialisation, so whilst the time is materially the same, it serialises to a different string

So, in that case it makes sense that the API might return ISO 8601 timestamps with a non-zero UTC offset in the GET appointments response. But in the contract tests I'm also seeing examples of non-zero offsets being supplied to the POST and PATCH endpoints too. And I'd like to understand what you need this offset for, and if that means there's an expectation on the client to calculate this offset?

I'd say that _if_ you do need time zone information, then a clearer way of communicating it is a tuple like (date, wall time, tz identifier), e.g. `('2021-03-01', '10:03:00.000', 'Europe/London')`.

File: server/services/interventionsService.ts (hunk 2 above)
Comment (tomsmyers) [reply server/services/interventionsService.ts:431]: you get some type warnings. it's an issue with the implementation of interfaces in typescript and this is the only way to get around it that was sensible

File: server/views/serviceProviderReferrals/showReferral.njk (hunk 4 above)
Comment (tomsmyers) [reply server/views/serviceProviderReferrals/showReferral.njk:29]: i cherry picked georges commit to fix this to make local testing palatable. does it matter which PR it's included in?

File: server/services/interventionsService.test.ts (hunk 5 above)
Comment (tomsmyers) [reply server/services/interventionsService.test.ts]: yup

File: server/views/serviceProviderReferrals/showReferral.njk (hunk 4 above)
Comment (lawrence-forooghian) [reply server/views/serviceProviderReferrals/showReferral.njk]: I think it's no longer necessary now that this is handled on the intervention progress page.

**Review by lawrence-forooghian — APPROVED** [review]
